File: servers/go/ws.go
package main

import (
	"bufio"
	"crypto/rand"
	"crypto/sha1"
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"os"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

const (
	wsGUID      = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
	maxFrameLen = 1 << 20 // 1 MiB inbound frame cap

	wsOpCont   = 0x0
	wsOpText   = 0x1
	wsOpBinary = 0x2
	wsOpClose  = 0x8
	wsOpPing   = 0x9
	wsOpPong   = 0xA

	pingInterval    = 20 * time.Second // engine.io v3 pingInterval advertised
	pingTimeout     = 15 * time.Second // engine.io v3 pingTimeout advertised
	readIdleTimeout = 90 * time.Second // hard cap on a fully silent TCP peer

	outboxSize = 512 // frames; slow consumers beyond this are dropped
)

// Conn is one engine.io v3 / socket.io v2 session over a hijacked TCP conn.
// player/lobbyID are guarded by World.mu; everything else by its own lifecycle.
type Conn struct {
	id       string
	nc       net.Conn
	br       *bufio.Reader
	out      chan []byte // complete socket.io/engine.io payloads as text frames
	ctl      chan ctlFrame
	dead     chan struct{}
	killOnce sync.Once
	wmu      sync.Mutex // serializes raw frame writes
	lastRecv atomic.Int64

	world *World

	player  *Player // guarded by world.mu
	lobbyID string  // guarded by world.mu
}

type ctlFrame struct {
	op   byte
	data []byte
}

func (c *Conn) touch() { c.lastRecv.Store(time.Now().UnixNano()) }

func (c *Conn) kill() {
	c.killOnce.Do(func() {
		close(c.dead)
		c.nc.Close()
	})
}

func (c *Conn) writeFrameSafe(op byte, payload []byte) error {
	c.wmu.Lock()
	defer c.wmu.Unlock()
	return writeWSFrame(c.nc, op, payload)
}

// trySendFrame enqueues a text frame; a full outbox marks the connection as a
// slow consumer and it is dropped (keeps memory bounded under load).
func trySendFrame(c *Conn, frame []byte) {
	select {
	case c.out <- frame:
	default:
		c.kill()
	}
}

func SendTo(c *Conn, event string, arg any) { trySendFrame(c, SioEvent(event, arg)) }

func BroadcastFrame(l *Lobby, frame []byte) {
	for _, m := range l.members {
		trySendFrame(m, frame)
	}
}

func SendFeed(l *Lobby, obj any) { BroadcastFrame(l, SioEvent("feed", obj)) }

// SioEvent builds a socket.io v2 EVENT frame: 42["name",arg].
func SioEvent(name string, arg any) []byte {
	b := marshalNoEsc([]any{name, arg})
	out := make([]byte, 0, len(b)+2)
	out = append(out, '4', '2')
	return append(out, b...)
}

func openPacket(sid string) []byte {
	return []byte(fmt.Sprintf(
		"0{\"sid\":\"%s\",\"upgrades\":[],\"pingInterval\":20000,\"pingTimeout\":15000}", sid))
}

// ---------------------------------------------------------------------------
// HTTP upgrade + session pumps.

func (s *Server) HandleSocketIO(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
		return
	}
	q := r.URL.Query()
	if q.Get("transport") != "websocket" {
		// Only the websocket transport is implemented (SPEC §Wire protocol).
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusBadRequest)
		w.Write([]byte(`{"code":3,"message":"unsupported transport"}`))
		return
	}
	if !strings.EqualFold(r.Header.Get("Upgrade"), "websocket") ||
		!connectionHasUpgrade(r.Header.Get("Connection")) {
		http.Error(w, "Bad Request", http.StatusBadRequest)
		return
	}
	key := r.Header.Get("Sec-WebSocket-Key")
	if key == "" {
		http.Error(w, "Bad Request", http.StatusBadRequest)
		return
	}
	hj, ok := w.(http.Hijacker)
	if !ok {
		http.Error(w, "Internal Server Error", http.StatusInternalServerError)
		return
	}
	nc, brw, err := hj.Hijack()
	if err != nil {
		return
	}
	resp := "HTTP/1.1 101 Switching Protocols\r\n" +
		"Upgrade: websocket\r\n" +
		"Connection: Upgrade\r\n" +
		"Sec-WebSocket-Accept: " + acceptKey(key) + "\r\n\r\n"
	if _, err := nc.Write([]byte(resp)); err != nil {
		nc.Close()
		return
	}
	if tc, ok := nc.(*net.TCPConn); ok {
		tc.SetNoDelay(true)
	}

	c := &Conn{
		id:    NewSID(),
		nc:    nc,
		br:    brw.Reader,
		out:   make(chan []byte, outboxSize),
		ctl:   make(chan ctlFrame, 16),
		dead:  make(chan struct{}),
		world: s.world,
	}
	c.touch()
	go c.writePump()
	go c.readPump()

	// engine.io open packet followed by the socket.io namespace CONNECT.
	trySendFrame(c, openPacket(c.id))
	trySendFrame(c, []byte("40"))
}

func connectionHasUpgrade(h string) bool {
	for _, part := range strings.Split(h, ",") {
		if strings.EqualFold(strings.TrimSpace(part), "Upgrade") {
			return true
		}
	}
	return false
}

func acceptKey(key string) string {
	h := sha1.New()
	h.Write([]byte(key + wsGUID))
	return base64.StdEncoding.EncodeToString(h.Sum(nil))
}

// NewSID returns an engine.io-style session id: 16 random bytes, base64url,
// i.e. 22 chars of [A-Za-z0-9_-].
func NewSID() string {
	var b [16]byte
	rand.Read(b[:])
	return base64.RawURLEncoding.EncodeToString(b[:])
}

func (c *Conn) writePump() {
	ticker := time.NewTicker(pingInterval)
	defer ticker.Stop()
	for {
		select {
		case <-c.dead:
			return
		case f := <-c.out:
			if err := c.writeFrameSafe(wsOpText, f); err != nil {
				c.kill()
				return
			}
		case cf := <-c.ctl:
			if err := c.writeFrameSafe(cf.op, cf.data); err != nil {
				c.kill()
				return
			}
		case <-ticker.C:
			age := time.Since(time.Unix(0, c.lastRecv.Load()))
			if age > pingInterval+pingTimeout {
				c.kill() // no pong within pingTimeout: close
				return
			}
			if err := c.writeFrameSafe(wsOpText, []byte("2")); err != nil {
				c.kill()
				return
			}
		}
	}
}

func (c *Conn) readPump() {
	defer func() {
		c.kill()
		c.world.Disconnect(c)
	}()

	var (
		fragOp byte
		frag   []byte
	)
	for {
		c.nc.SetReadDeadline(time.Now().Add(readIdleTimeout))
		op, data, fin, err := readWSFrame(c.br)
		if err != nil {
			return
		}
		if os.Getenv("SNEK_WS_DEBUG") == "1" {
		}
		switch op {
		case wsOpClose:
			echo := data
			if len(echo) > 125 {
				echo = echo[:125]
			}
			_ = c.writeFrameSafe(wsOpClose, echo) // RFC: echo close, then go away
			return
		case wsOpPing:
			c.touch()
			select {
			case c.ctl <- ctlFrame{op: wsOpPong, data: data}:
			default:
				c.kill()
				return
			}
		case wsOpPong:
			c.touch() // resets the heartbeat timer per SPEC
		case wsOpBinary:
			c.touch() // tolerated, ignored
		case wsOpText:
			c.touch()
			if fin {
				c.handleMessage(data)
			} else {
				fragOp, frag = op, append([]byte(nil), data...)
			}
		case wsOpCont:
			if fragOp == 0 {
				return // continuation without a started message
			}
			frag = append(frag, data...)
			if fin {
				msg := frag
				fragOp, frag = 0, nil
				c.handleMessage(msg)
			}
		default:
			return
		}
	}
}

// handleMessage dispatches engine.io packets: 2 ping, 3 pong, 4 message.
func (c *Conn) handleMessage(b []byte) {
	if os.Getenv("SNEK_WS_DEBUG") == "1" {
	}
	if len(b) == 0 {
		return
	}
	switch b[0] {
	case '2': // client-initiated engine ping -> pong
		trySendFrame(c, []byte("3"))
	case '3': // engine pong (heartbeat already recorded via touch)
	case '4':
		c.handleSIO(b[1:])
	}
}

// handleSIO dispatches socket.io v2 packets carried in engine messages:
// 40 CONNECT, 41 DISCONNECT, 42 EVENT.
func (c *Conn) handleSIO(b []byte) {
	if len(b) == 0 {
		return
	}
	switch b[0] {
	case '0':
		// Namespace CONNECT from the client; the default namespace was
		// connected server-side already, nothing to do.
	case '1':
		c.kill() // namespace disconnect == end of session here
	case '2':
		// EVENT. Must be 42[...json array...]; other namespaces ignored.
		if len(b) < 3 || b[1] != '[' {
			return
		}
		var ev []any
		if err := json.Unmarshal(b[1:], &ev); err != nil { // JSON array follows the type digit
			return
		}
		if len(ev) == 0 {
			return
		}
		name, _ := ev[0].(string)
		args := ev[min(1, len(ev)):]
		switch name {
		case "clientReady":
			c.world.ClientReady(c, args)
		case "keyPress":
			var data any
			if len(args) > 0 {
				data = args[0]
			}
			c.world.KeyPress(c, data)
		}
	}
}

// ---------------------------------------------------------------------------
// WebSocket framing (RFC 6455 subset sufficient for engine.io v3).

// readWSFrame reads one frame, unmasks client payload, and reports FIN.
func readWSFrame(br *bufio.Reader) (opcode byte, payload []byte, fin bool, err error) {
	var hdr [8]byte
	if _, err = ioReadFull(br, hdr[:2]); err != nil {
		return
	}
	fin = hdr[0]&0x80 != 0
	opcode = hdr[0] & 0x0f
	masked := hdr[1]&0x80 != 0
	length := uint64(hdr[1] & 0x7f)
	switch length {
	case 126:
		if _, err = ioReadFull(br, hdr[:2]); err != nil {
			return
		}
		length = uint64(binary.BigEndian.Uint16(hdr[:2]))
	case 127:
		if _, err = ioReadFull(br, hdr[:8]); err != nil {
			return
		}
		length = binary.BigEndian.Uint64(hdr[:8])
		if length&(1<<63) != 0 {
			err = errProtocol
			return
		}
	}
	if length > maxFrameLen {
		err = errTooLarge
		return
	}
	var mk []byte
	if masked {
		mk = make([]byte, 4)
		if _, err = ioReadFull(br, mk); err != nil { // masking key precedes payload (RFC 6455 §5.3)
			return
		}
	}
	payload = make([]byte, length)
	if _, err = ioReadFull(br, payload); err != nil {
		return
	}
	if masked {
		for i := range payload {
			payload[i] ^= mk[i&3]
		}
	}
	return
}

var errProtocol = fmt.Errorf("websocket protocol error")
var errTooLarge = fmt.Errorf("frame too large")

func ioReadFull(r *bufio.Reader, buf []byte) (int, error) {
	n := 0
	for n < len(buf) {
		nn, err := r.Read(buf[n:])
		n += nn
		if err != nil {
			return n, err
		}
	}
	return n, nil
}

// writeWSFrame writes one unmasked server frame (single Write per frame).
func writeWSFrame(w net.Conn, op byte, payload []byte) error {
	n := len(payload)
	hdr := make([]byte, 0, 10)
	hdr = append(hdr, 0x80|op)
	switch {
	case n < 126:
		hdr = append(hdr, byte(n))
	case n < 65536:
		hdr = append(hdr, 126, byte(n>>8), byte(n))
	default:
		ext := make([]byte, 8)
		binary.BigEndian.PutUint64(ext, uint64(n))
		hdr = append(hdr, 127)
		hdr = append(hdr, ext...)
	}
	buf := make([]byte, 0, len(hdr)+n)
	buf = append(buf, hdr...)
	buf = append(buf, payload...)
	_, err := w.Write(buf)
	return err
}

File: servers/rust/src/ws.rs
//! WebSocket (RFC 6455) server-side codec plus the engine.io v3 /
//! socket.io v2 session layer, per SPEC wire protocol section.

use std::collections::HashMap;
use std::io::{BufRead, Write};
use std::net::{Shutdown, TcpStream};
use std::sync::mpsc::{channel, RecvTimeoutError, Sender};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use crate::base64;
use crate::http::Req;
use crate::json::Json;
use crate::rng;
use crate::sha1;

pub const PING_INTERVAL_MS: u64 = 20_000;
pub const PING_TIMEOUT_MS: u64 = 15_000;
const MAX_FRAME_BYTES: usize = 8 * 1024 * 1024;
const WS_GUID: &str = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";

/// Messages sent from readers/game threads to a connection writer thread.
pub enum Out {
    /// Full wire payload of an engine.io text packet, e.g. "42[\"init\",...]".
    Text(String),
    /// Reader saw an engine.io pong ('3') - reset the ping timeout.
    GotPong,
    /// Reader saw an engine.io ping ('2') - reply with '3'.
    ReplyPong,
    /// Reply to a WebSocket-level ping.
    WsPong(Vec<u8>),
    /// Echo the peer close code and finish.
    CloseEcho(Vec<u8>),
    Close,
}

type Registry = Arc<Mutex<HashMap<String, Sender<Out>>>>;

fn registry() -> Registry {
    static REG: std::sync::OnceLock<Registry> = std::sync::OnceLock::new();
    REG.get_or_init(|| Arc::new(Mutex::new(HashMap::new())))
        .clone()
}

fn game() -> &'static Mutex<crate::game::GameServer> {
    static G: std::sync::OnceLock<Mutex<crate::game::GameServer>> = std::sync::OnceLock::new();
    G.get_or_init(|| Mutex::new(crate::game::GameServer::new()))
}

fn dispatch(out: Vec<crate::game::OutMsg>, reg: &Registry) {
    if out.is_empty() {
        return;
    }
    let reg = reg.lock().unwrap_or_else(|e| e.into_inner());
    for (sid, payload) in out {
        if let Some(tx) = reg.get(&sid) {
            let _ = tx.send(Out::Text(payload));
        }
    }
}

pub fn with_game<T>(f: impl FnOnce(&mut crate::game::GameServer) -> T) -> T {
    let mut g = game().lock().unwrap_or_else(|e| e.into_inner());
    f(&mut g)
}

fn compute_accept(key: &str) -> String {
    base64::encode_std(&sha1::sha1(format!("{}{}", key, WS_GUID).as_bytes()))
}

// ------------------------------ frame codec ---------------------------------

struct Frame {
    opcode: u8,
    fin: bool,
    payload: Vec<u8>,
}

fn read_frame(r: &mut dyn BufRead) -> std::io::Result<Option<Frame>> {
    let mut hdr = [0u8; 2];
    match r.read_exact(&mut hdr) {
        Ok(_) => {}
        Err(e) => return Err(e),
    }
    let fin = hdr[0] & 0x80 != 0;
    let rsv = hdr[0] & 0x70;
    let opcode = hdr[0] & 0x0F;
    let masked = hdr[1] & 0x80 != 0;
    let mut len = (hdr[1] & 0x7F) as usize;
    if rsv != 0 {
        return Err(std::io::Error::new(
            std::io::ErrorKind::InvalidData,
            "rsv bits set",
        ));
    }
    if len == 126 {
        let mut ext = [0u8; 2];
        r.read_exact(&mut ext)?;
        len = u16::from_be_bytes(ext) as usize;
    } else if len == 127 {
        let mut ext = [0u8; 8];
        r.read_exact(&mut ext)?;
        len = u64::from_be_bytes(ext) as usize;
    }
    if len > MAX_FRAME_BYTES {
        return Err(std::io::Error::new(
            std::io::ErrorKind::InvalidData,
            "frame too large",
        ));
    }
    let mask = if masked {
        let mut m = [0u8; 4];
        r.read_exact(&mut m)?;
        Some(m)
    } else {
        None
    };
    let mut payload = vec![0u8; len];
    if len > 0 {
        r.read_exact(&mut payload)?;
    }
    if let Some(m) = mask {
        for (i, b) in payload.iter_mut().enumerate() {
            *b ^= m[i % 4];
        }
    }
    Ok(Some(Frame {
        opcode,
        fin,
        payload,
    }))
}

fn write_frame(w: &mut TcpStream, opcode: u8, payload: &[u8]) -> std::io::Result<()> {
    let mut head = Vec::with_capacity(10);
    head.push(0x80 | opcode);
    let len = payload.len();
    if len < 126 {
        head.push(len as u8);
    } else if len <= u16::MAX as usize {
        head.push(126);
        head.extend_from_slice(&(len as u16).to_be_bytes());
    } else {
        head.push(127);
        head.extend_from_slice(&(len as u64).to_be_bytes());
    }
    w.write_all(&head)?;
    w.write_all(payload)?;
    w.flush()
}

// ------------------------------ session -------------------------------------

/// Handle a GET /socket.io/?EIO=3&transport=websocket upgrade for this
/// connection. Takes ownership of the stream; returns when the socket dies.
pub fn run_session(mut sock: TcpStream, reader: &mut dyn BufRead, req: &Req) {
    let key = match req.header("sec-websocket-key") {
        Some(k) if !k.is_empty() => k.to_string(),
        _ => return,
    };
    let accept = compute_accept(&key);
    let resp = format!(
        "HTTP/1.1 101 Switching Protocols\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Accept: {}\r\n\r\n",
        accept
    );
    if sock.write_all(resp.as_bytes()).is_err() {
        return;
    }

    let mut writer_sock = match sock.try_clone() {
        Ok(s) => s,
        Err(_) => return,
    };

    let sid = {
        let mut r = rng::global().lock().unwrap_or_else(|e| e.into_inner());
        r.sid()
    };
    let (tx, rx) = channel::<Out>();

    // Writer thread: the ONLY place frames are written once the session runs.
    let writer = std::thread::Builder::new()
        .name("ws-writer".into())
        .stack_size(128 * 1024)
        .spawn(move || {
            // Engine.io v3 semantics (matching the reference stack): the
            // client pings every interval and ANY inbound packet proves
            // liveness; a silent peer is closed after interval+timeout.
            // We additionally emit our own "2" every interval (SPEC §Wire).
            let mut last_ping = Instant::now();
            let mut last_seen = Instant::now();
            loop {
                let now = Instant::now();
                if now.duration_since(last_ping) >= Duration::from_millis(PING_INTERVAL_MS) {
                    if write_frame(&mut writer_sock, 0x1, b"2").is_err() {
                        break;
                    }
                    last_ping = now;
                }
                if now.duration_since(last_seen)
                    >= Duration::from_millis(PING_INTERVAL_MS + PING_TIMEOUT_MS)
                {
                    let _ = write_frame(&mut writer_sock, 0x8, &[0x03, 0xE8]);
                    break;
                }
                match rx.recv_timeout(Duration::from_millis(50)) {
                    Ok(Out::Text(payload)) => {
                        if write_frame(&mut writer_sock, 0x1, payload.as_bytes()).is_err() {
                            break;
                        }
                    }
                    Ok(Out::GotPong) => last_seen = Instant::now(),
                    Ok(Out::ReplyPong) => {
                        if write_frame(&mut writer_sock, 0x1, b"3").is_err() {
                            break;
                        }
                    }
                    Ok(Out::WsPong(p)) => {
                        if write_frame(&mut writer_sock, 0xA, &p).is_err() {
                            break;
                        }
                    }
                    Ok(Out::CloseEcho(code)) => {
                        let _ = write_frame(&mut writer_sock, 0x8, &code);
                        break;
                    }
                    Ok(Out::Close) | Err(RecvTimeoutError::Disconnected) => {
                        let _ = write_frame(&mut writer_sock, 0x8, &[0x03, 0xE8]);
                        break;
                    }
                    Err(RecvTimeoutError::Timeout) => continue,
                }
            }
            let _ = writer_sock.shutdown(Shutdown::Both);
        });
    if writer.is_err() {
        return;
    }

    // Register before anything can be sent to this sid.
    registry()
        .lock()
        .unwrap_or_else(|e| e.into_inner())
        .insert(sid.clone(), tx.clone());

    // Engine.io open packet, then the socket.io CONNECT frame.
    let open_packet = format!(
        "0{{\"sid\":\"{}\",\"upgrades\":[],\"pingInterval\":{},\"pingTimeout\":{}}}",
        sid, PING_INTERVAL_MS, PING_TIMEOUT_MS
    );
    let _ = tx.send(Out::Text(open_packet));
    let _ = tx.send(Out::Text("40".to_string()));

    // ---- reader loop (this thread) ----
    let mut frag: Vec<u8> = Vec::new();
    'read: while let Ok(Some(frame)) = read_frame(reader) {
        match frame.opcode {
            0x0 => {
                frag.extend_from_slice(&frame.payload);
                if frame.fin {
                    let text = String::from_utf8_lossy(&frag).into_owned();
                    frag.clear();
                    if handle_text(&tx, &sid, &text) == Flow::Stop {
                        break 'read;
                    }
                } else if frag.len() > MAX_FRAME_BYTES {
                    break 'read;
                }
            }
            0x1 | 0x2 => {
                let text = String::from_utf8_lossy(&frame.payload).into_owned();
                if frame.fin {
                    if handle_text(&tx, &sid, &text) == Flow::Stop {
                        break 'read;
                    }
                } else {
                    frag = text.into_bytes();
                }
            }
            0x8 => {
                let code: Vec<u8> = if frame.payload.len() >= 2 {
                    frame.payload[..2].to_vec()
                } else {
                    vec![0x03, 0xE8]
                };
                let _ = tx.send(Out::CloseEcho(code));
                break 'read;
            }
            0x9 => {
                let _ = tx.send(Out::WsPong(frame.payload));
            }
            0xA => {}
            _ => {}
        }
    }

    // ---- cleanup ----
    registry()
        .lock()
        .unwrap_or_else(|e| e.into_inner())
        .remove(&sid);
    with_game(|g| g.mark_closed(&sid));
    let _ = tx.send(Out::Close);
    drop(tx);
    let _ = sock.shutdown(Shutdown::Both); // wakes the blocked read
}

#[derive(PartialEq)]
enum Flow {
    Continue,
    Stop,
}

/// Process one engine.io text packet. Returns Stop when the session must end.
fn handle_text(tx: &Sender<Out>, sid: &str, text: &str) -> Flow {
    let Some(first) = text.bytes().next() else {
        return Flow::Continue;
    };
    // Any inbound engine.io packet is a sign of life (engine.io v3 server
    // behaviour). The reply mapping below follows the packet type.
    let _ = tx.send(Out::GotPong);
    match first {
        b'2' => {
            let _ = tx.send(Out::ReplyPong);
            Flow::Continue
        }
        b'3' => {
            let _ = tx.send(Out::GotPong);
            Flow::Continue
        }
        b'4' => handle_message(sid, &text[1..]),
        _ => Flow::Continue,
    }
}

fn handle_message(sid: &str, rest: &str) -> Flow {
    let Some(second) = rest.bytes().next() else {
        return Flow::Continue;
    };
    match second {
        b'0' => Flow::Continue, // client nsp CONNECT; we already answered
        b'1' => Flow::Stop,     // client nsp DISCONNECT: treat the session as over
        b'2' => {
            // Event: 42["name",arg,...]; tolerate an ack id between 42 and [.
            let payload = &rest[1..];
            let Some(pos) = payload.find('[') else {
                return Flow::Continue;
            };
            let arr = match crate::json::parse(&payload[pos..]) {
                Ok(Json::Arr(a)) => a,
                _ => return Flow::Continue,
            };
            let Some(name) = arr.first().and_then(|v| v.as_str()) else {
                return Flow::Continue;
            };
            let args: &[Json] = if arr.len() > 1 { &arr[1..] } else { &[] };
            let mut out: Vec<crate::game::OutMsg> = Vec::new();
            with_game(|g| g.handle_event(sid, name, args, &mut out));
            dispatch(out, &registry());
            Flow::Continue
        }
        _ => Flow::Continue,
    }
}

// --------------------- handles for main.rs / ticker --------------------------

pub type RegistryHandle = Registry;

pub fn registry_handle() -> RegistryHandle {
    registry().clone()
}

pub fn dispatch_handle(out: Vec<crate::game::OutMsg>, reg: &RegistryHandle) {
    dispatch(out, reg)
}
